sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,214 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_spi_device.h"
+
+#include "spi_device_regs.h"  // Generated.
+
+// This is a hack to get the SPI offsets out of regs.h.
+// DO NOT MERGE.
+#define SPI_DEVICE0_BASE_ADDR 0x0
+#define SPI_DEVICE_CONTROL_REG SPI_DEVICE_CONTROL(0)
+#define SPI_DEVICE_CFG_REG SPI_DEVICE_CFG(0)
+#define SPI_DEVICE_RXF_ADDR_REG SPI_DEVICE_RXF_ADDR(0)
+#define SPI_DEVICE_TXF_ADDR_REG SPI_DEVICE_TXF_ADDR(0)
+
+#define SPI_DEVICE_RXF_PTR_REG SPI_DEVICE_RXF_PTR(0)
+#define SPI_DEVICE_TXF_PTR_REG SPI_DEVICE_TXF_PTR(0)
+
+#define SPI_DEVICE_BUFFER_REG SPI_DEVICE_BUFFER(0)
+
+#define SPI_DEVICE_RXF_OFFSET(ptr)               \
+  (SPI_DEVICE_BUFFER_REG + SPI_DEVICE_RXF_BASE + \
+   ((ptr) & (SPI_DEVICE_RXF_SIZE - 1) & ~0b11))
+#define SPI_DEVICE_TXF_OFFSET(ptr)               \
+  (SPI_DEVICE_BUFFER_REG + SPI_DEVICE_TXF_BASE + \
+   ((ptr) & (SPI_DEVICE_TXF_SIZE - 1) & ~0b11))
+
+#define SPI_DEVICE_RXF_BASE 0x000
+#define SPI_DEVICE_RXF_SIZE 0x400
+#define SPI_DEVICE_TXF_BASE 0x600
+#define SPI_DEVICE_TXF_SIZE 0x200
+
+#define PHASE_MASK (SPI_DEVICE_BUFFER_SIZE_BYTES)
+#define VALUE_MASK (SPI_DEVICE_BUFFER_SIZE_BYTES - 1)
+
+void spi_device_init(reg32_t base_addr) {
+  // Reset the abort bit.
+  reg32_clear_bit(base_addr, SPI_DEVICE_CONTROL_REG, SPI_DEVICE_CONTROL_ABORT);
+
+  // The reset values for all SPI_DEVICE.CFG bits is 0, except for |timer_v|,
+  // which is all-ones (i.e., 0x7f).
+  uint32_t device_config = (0x7f & SPI_DEVICE_CFG_TIMER_V_MASK)
+                           << SPI_DEVICE_CFG_TIMER_V_OFFSET;
+  reg32_write(base_addr, SPI_DEVICE_CFG_REG, device_config);
+
+  uint32_t rxf_addr =
+      ((SPI_DEVICE_RXF_BASE & SPI_DEVICE_RXF_ADDR_BASE_MASK)
+       << SPI_DEVICE_RXF_ADDR_BASE_OFFSET) |
+      (((SPI_DEVICE_RXF_SIZE - 1) & SPI_DEVICE_RXF_ADDR_LIMIT_MASK)
+       << SPI_DEVICE_RXF_ADDR_LIMIT_OFFSET);
+  reg32_write(base_addr, SPI_DEVICE_RXF_ADDR_REG, rxf_addr);
+
+  uint32_t txf_addr =
+      ((SPI_DEVICE_TXF_BASE & SPI_DEVICE_TXF_ADDR_BASE_MASK)
+       << SPI_DEVICE_TXF_ADDR_BASE_OFFSET) |
+      (((SPI_DEVICE_TXF_SIZE - 1) & SPI_DEVICE_TXF_ADDR_LIMIT_MASK)
+       << SPI_DEVICE_TXF_ADDR_LIMIT_OFFSET);
+  reg32_write(base_addr, SPI_DEVICE_TXF_ADDR_REG, txf_addr);
+}
+
+static inline size_t fifo_depth(uint32_t write_addr, uint32_t read_addr,
```

Comment:
It seems to me that the function is a bit too big for inline, the same with the other inline functions.

Also, I know that often static functions are not documented, but in a spirit of a DIF document that (from my understanding) strives for simplicity, readability and being "user" friendly - it is a good idea to document the static functions as well.

Reply:
I disagree, these functions are totally reasonable as inline functions.

If we care about binary size, `-Os` and `-Oz` can cause the compiler to not inline them if it decides that's a good idea: after all, `inline` is only a hint.

Reply:
That's fair enough, I haven't got a strong opinion on this, just decided to flag it up just in case.

Reply:
I think we should in the DIF specification establish criterias for inlining functions and set some rules